File: services/mcpserver/src/forensics_mcp/chain.py
"""Recompute the decision-log hash chain. This MUST match the Go writer's canonical form byte for
byte (services/api/internal/chain/chain.go): hash = SHA-256(prev_hash || "seq|action|lawful_basis"
|| subject_hash), genesis prev_hash = 32 zero bytes.
"""

from __future__ import annotations

import hashlib

GENESIS_PREV_HASH = b"\x00" * 32


def link(prev_hash: bytes, seq: int, action: str, lawful_basis: str, subject_hash: bytes) -> bytes:
    canonical = f"{seq}|{action}|{lawful_basis}".encode() + subject_hash
    h = hashlib.sha256()
    h.update(bytes(prev_hash))
    h.update(canonical)
    return h.digest()

File: services/api/internal/chain/chain.go
// Package chain computes the decision-log hash chain. The Python forensics verifier recomputes the
// identical bytes, so the canonical form defined here is a cross-language contract: keep it and the
// verifier in lockstep.
package chain

import (
	"crypto/sha256"
	"fmt"
)

// GenesisPrevHash returns the prev_hash of the first decision-log row (32 zero bytes). It returns
// a fresh slice each call so no caller can mutate a shared package-global.
func GenesisPrevHash() []byte { return make([]byte, 32) }

// SubjectHash pseudonymizes a subject id for the decision log: SHA-256 over the id's string form.
// No raw personal identifier ever reaches the retained log.
func SubjectHash(subjectID string) []byte {
	h := sha256.Sum256([]byte(subjectID))
	return h[:]
}

// canonicalRow is the deterministic byte encoding of a decision-log row's content that is hashed
// into the chain: "seq|action|lawful_basis" followed by the raw subject-hash bytes.
func canonicalRow(seq int64, action, lawfulBasis string, subjectHash []byte) []byte {
	prefix := []byte(fmt.Sprintf("%d|%s|%s", seq, action, lawfulBasis))
	return append(prefix, subjectHash...)
}

// Link computes hash = SHA-256(prevHash || canonicalRow) for a new decision-log row.
func Link(prevHash []byte, seq int64, action, lawfulBasis string, subjectHash []byte) []byte {
	h := sha256.New()
	h.Write(prevHash)
	h.Write(canonicalRow(seq, action, lawfulBasis, subjectHash))
	return h.Sum(nil)
}
